### PR TITLE
Markdown

### PR DESCRIPTION
**Pull request by emord**
This copies the help text to the label xml under "markdown"

More info:

https://docs.google.com/a/dimagi.com/document/d/1FsV7I-08psf5zbAC3Hk8M9IZ4a9JfHOvSZrMjeE9jE4/edit#
https://confluence.dimagi.com/display/TDH/Markdown

Not the most clear code, but didn't see a great way to do this in javarosa.

### CODE REVIEW

**Comment by emord** [comment]
review buddy @snopoke

**Comment by snopoke** [comment]
Looks good on the surface but don't have enough context to merge. Will let @millerdev merge.

**Comment by emord** [comment]
Will add a feature flag for TDH before releasing

File: src/javaRosa.js [hunk 1]
```diff
@@ -310,8 +310,9 @@ define([
         updateForMug: function (mug, defaultLabelValue) {
             // set default itext id/values
             if (!mug.options.isDataOnly) {
+                var item;
```
Comment (millerdev) [inline src/javaRosa.js]: What do you think of renaming this variable to `label` now that it's scope reaches across the function to the help itext item as well?

File: src/javaRosa.js [hunk 2]
```diff
@@ -319,6 +320,9 @@ define([
                 }
                 if (!mug.p.helpItext && mug.spec.helpItext.presence !== "notallowed") {
                     mug.p.helpItext = this.createItem();
+                    if (item) {
+                        item.set("", 'markdown');
```
Comment (millerdev) [inline src/javaRosa.js]: This seems odd to me, but I may be missing something. It's clearing the content of the "markdown" form for all languages in the label itext item when creating the help itext item. Is that correct?

File: tests/javaRosa.js [hunk 3]
```diff
@@ -785,6 +785,7 @@ require([
                     <translation lang="en" default="">\
                         <text id="question1-label">\
                             <value>question1 en label</value>\
+                            <value form="markdown">question1 en help</value>\
```
Comment (millerdev) [inline tests/javaRosa.js]: Why are we putting the help-text markdown content in the label itext item instead of in the help itext item? Where will we put label-text markdown content if at some point in the future we want to allow markdown in the label as well?

Comment (emord) [inline tests/javaRosa.js]: I misread the spec. Confirmed with Ro and will. Should be in question1-help now.

File: src/javaRosa.js [hunk 4]
```diff
@@ -1760,13 +1768,28 @@ define([
...
+                                return;
+                            }
+                            var mug = this.mug,
+                                helpItext = mug.p.helpItext,
+                                helpItextForm = helpItext.forms[0],
+                                markdownForms = _.find(helpItext.forms, function(itext) {
+                                    return itext.name === 'markdown';
+                                });
+                            if (markdownForms) {
+                                markdownForms.data = JSON.parse(JSON.stringify(helpItextForm.data));
```
Comment (millerdev) [inline src/javaRosa.js]: Would `_.clone(helpItextForm.data)` be appropriate here? That's how it's done in [`ItextForm.clone()`](https://github.com/dimagi/Vellum/blob/markdown/src/javaRosa.js#L176)

Comment (emord) [inline src/javaRosa.js]: Possibly. This creates a deep copy where as `_.clone` does a shallow copy.

Comment (millerdev) [inline src/javaRosa.js]: Yep. If that matters here, then it probably matters in `ItextForm.clone()` too, but I don't think it does. AFAIK, `data` is a mapping of language codes to itext content strings, so a shallow copy should be fine. Happy to hear if you can think of a reason that it matters.

Comment (emord) [inline src/javaRosa.js]: Yeah. I double checked and shallow copies are ok

File: tests/static/all_question_types.xml [hunk 5]
```diff
@@ -85,6 +85,7 @@
 					</text>
 					<text id="question1-help">
 						<value>question1 en help</value>
+						<value form="markdown">question1 en help</value>
```
Comment (millerdev) [inline tests/static/all_question_types.xml]: Would it make sense to omit the markdown itext form if it's identical to the non-markdown version? Would that add a lot of complexity? Seems like wasted bandwidth. Might want to check with @ctsims on this.

Comment (emord) [inline tests/static/all_question_types.xml]: I think @wpride worked on this. He can probably comment on it.

Comment (wpride) [inline tests/static/all_question_types.xml]: I think that would be fine, the phone behavior would be the same either way. The only edge case I can think of is if you were trying to use markdown but the plain text also had markdown in it making it identical so Vellum eats the markdown version even though you wanted that... but even typing that out that just seems completely ridiculous that you'd want plain text with \n and # riddled all over it so, yeah, fine

If the "markdown" form exists it will always be used instead of the normal version FWIW

Comment (emord) [inline tests/static/all_question_types.xml]: The spec says: 
> automatically treat all text in the ‘help’ text field as markdown by generating the xml for the xforms with duplicate <value form=”markdown”></value> fields

So I think that means it's intended for markdown & non-markdown text will be the same in all cases for now. For this release there isn't any pretty UI for making the markdown as it's only for TDH.

If the feature flag is set does that mean that we would only want to write the markdown form? I'm not sure how commcare would react if it was given <value form="markdown"> but not <value>

Comment (millerdev) [inline tests/static/all_question_types.xml]: Hmm, so we're duplicating the `<value>` content in `<value form="markdown">` as a flag to say treat this text as markdown. That design seems less than ideal. Is there a way for the phone to access the flag that says "treat all help text as markdown for this form"?

Comment (wpride) [inline tests/static/all_question_types.xml]: There is not - we opted against that design

Comment (millerdev) [inline tests/static/all_question_types.xml]: Fair enough. @emord are you ready to have this merged?

Comment (emord) [inline tests/static/all_question_types.xml]: Yeah. Should be good to go.
